csrc/executor_utils.cpp

```diff
@@ -407,41 +407,113 @@ void validateAlignedVectorizeExtents(
   //     info.word_size);
 }
 
+namespace {
+
+// Return offsets of the first points accessed as well as slcied root
+// domains. Currently only non-zero when tensor is sliced.
+std::pair<std::unordered_set<size_t>, std::unordered_set<IterDomain*>>
+getTensorOffsets(
+    TensorView* tv,
+    const std::vector<int64_t>& logical_strides,
+    ExpressionEvaluator& eval) {
+  if (!tv->isFusionInput()) {
+    return {{0}, {}};
+  }
+
+  std::unordered_set<size_t> offsets;
+  std::unordered_set<IterDomain*> sliced_domains;
+
+  for (auto use : tv->uses()) {
+    auto slice = dynamic_cast<SliceOp*>(use);
+
+    if (slice == nullptr) {
+      offsets.insert(0);
+      continue;
+    }
+
+    const auto root_ids =
+        TensorDomain::noReductions(tv->getMaybeRFactorDomain());
+    TORCH_INTERNAL_ASSERT(logical_strides.size() == root_ids.size());
+    const auto slice_info = slice->getRanges();
+
+    size_t offset = 0;
+    for (const auto i : c10::irange(root_ids.size())) {
+      auto slice_start_eval = eval.evaluate(slice_info.at(i).start);
+      TORCH_INTERNAL_ASSERT(slice_start_eval.hasValue());
+      auto slice_stop_eval = eval.evaluate(slice_info.at(i).stop);
+      TORCH_INTERNAL_ASSERT(slice_stop_eval.hasValue());
+      auto extent_eval =
+          eval.evaluate(root_ids.at(i)->getMaybeExpandedExtent());
+      TORCH_INTERNAL_ASSERT(extent_eval.hasValue());
+
+      offset += static_cast<size_t>(
+          slice_start_eval.as<int64_t>() * logical_strides.at(i));
+
+      // Keep track of the root domain unless this slice is
+      // effectively no-op
+      if (slice_start_eval.as<int64_t>() != 0 ||
+          slice_stop_eval.as<int64_t>() != extent_eval.as<int64_t>()) {
+        sliced_domains.insert(root_ids.at(i));
+      }
+    }
+
+    offsets.insert(offset);
+  }
+
+  return std::make_pair(offsets, sliced_domains);
+}
+
+} // namespace
+
 void validateAlignedVectorizedFusionInputOutput(
     const at::Tensor& aten_tensor,
     int word_size,
     TensorView* tv,
-    ExpressionEvaluator eval) {
+    ExpressionEvaluator& eval) {
   eval.bind(tv, aten_tensor);
   auto metadata = eval.evaluate(IrBuilder::metadataExpr(tv));
 
+  const auto [offsets, sliced_domains] = getTensorOffsets(
+      tv, std::vector<int64_t>(metadata["logical_stride"]), eval);
+  const bool is_sliced = !sliced_domains.empty();
+
+  const auto& domain_to_validate =
+      is_sliced ? tv->getMaybeRFactorDomain() : tv->getMaybeAllocationDomain();
```

Comment:
@zasdfgbnm Validation of strides and contiguity is usually done with the allocation domain, but for sliced tensors, it also depends on whether an rfactor domain is sliced or not, so for such tensors it seems we need to look at the rfactor domain. Does this make sense? It seems reasonable to me, but I may be missing something.